test/addons-napi/test_promise/test_promise.c

```diff
@@ -0,0 +1,65 @@
+#include <node_api.h>
+#include "../common.h"
+
+napi_ref deferred_ref = NULL;
+
+napi_value createPromise(napi_env env, napi_callback_info info) {
+  napi_value promise, deferred, in_one_call;
+  bool one_call;
+  size_t argc = 1;
+
+  NAPI_CALL(env, napi_get_cb_info(env, info, &argc, &in_one_call, NULL, NULL));
+  NAPI_CALL(env, napi_get_value_bool(env, in_one_call, &one_call));
+
+  if (deferred_ref != NULL) {
+    NAPI_CALL(env, napi_delete_reference(env, deferred_ref));
+  }
+
+  if (one_call) {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, &promise));
+  } else {
+    NAPI_CALL(env, napi_create_promise(env, &deferred, NULL));
+    NAPI_CALL(env, napi_get_promise_from_deferred(env, deferred, &promise));
+  }
+  NAPI_CALL(env, napi_create_reference(env, deferred, 1, &deferred_ref));
+
+  return promise;
+}
```

Comment:
It does seem like we'd want to prevent the deferred object from being used as promise as I don't think we can ensure that's the case for all  js implementations.  I think it will just be too easy for an impl to pass back a deferred and then use it as a promise, and if it works we'll have code that breaks across different runtimes.  

We use this for wrap/unwrap:

```
  // Search the object's prototype chain for the wrapper with an internal field.
  // Usually the wrapper would be the first in the chain, but it is OK for
  // other objects to be inserted in the prototype chain.
  v8::Local<v8::Object> wrapper = obj;
  do {
    v8::Local<v8::Value> proto = wrapper->GetPrototype();
    RETURN_STATUS_IF_FALSE(
      env, !proto.IsEmpty() && proto->IsObject(), napi_invalid_arg);
    wrapper = proto.As<v8::Object>();
  } while (wrapper->InternalFieldCount() != 1);

  v8::Local<v8::Value> unwrappedValue = wrapper->GetInternalField(0);
  RETURN_STATUS_IF_FALSE(env, unwrappedValue->IsExternal(), napi_invalid_arg);

  *result = unwrappedValue.As<v8::External>()->Value();
```

So maybe we should generalize the code and use it for wrap and for creating a deferred object where the resolver is hidden in an internal field of the prototype chain ?